pytestqt/plugin.py

```diff
@@ -368,12 +372,25 @@ def wait(self):
         if self.timeout is None and not self._signals:
             raise ValueError("No signals or timeout specified.")
         if self.timeout is not None:
-            QtCore.QTimer.singleShot(self.timeout, self._loop.quit)
+            self._timer.timeout.connect(self._quit_loop_by_timeout)
+            self._timer.start()
         self._loop.exec_()
         if not self.signal_triggered and self.raising:
             raise SignalTimeoutError("Didn't get signal after %sms." %
                                       self.timeout)
 
+    def _quit_loop_by_timeout(self):
+        self._loop.quit()
+        self._cleanup()
```

Comment:
Shouldn't cleanup **always** be called? If so, I think you should call it as soon after `self._loop.exec_()` finishes (in line 377).